HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,32 @@
+def bfs(SOURCE,Graph,sc):
+    #set source as first node at distance 0    
+    node1 = [(SOURCE,0)]
+    node1 = sc.parallelize(node1).partitionBy(8)
+    final_rdd1 = node1
+    i = 0
+    
+    #initialize an accumulator and set to 0. Create while loop to run
+    #while accumulator matches value of the i iterator 
+    accumulator = sc.accumulator(0)
+    i = 0
+    while accumulator.value == i:        
+        #find neighbors of input node        
+        node1 = Graph.join(node1).distinct().values().partitionBy(8) 
+        
+        #new node only includes neighbors not already searched/in Final RDD        
+        node1 = node1.mapValues(lambda x: x+1).subtractByKey(final_rdd1)
```

Comment:
Instead of comparing to the original RDD each time, one approach (and the one that I took) would be to color each node either WHITE, BLACK, or GRAY, standing for - "already been searched, no new information", "hasn't been searched yet", and "going to step from this node on this iteration". You increment the accumulator every time you change a node to GRAY. Then you stop when there are no more GRAY nodes.